### PR TITLE
Always pass execution space to `Impl::DeepCopy`

### PR DESCRIPTION
**Pull request by pzehner**
This PR is a refactor which is part of a greater effort to simplify the use of `Impl::DeepCopy`.

Follow up of #7518.

It uses the 4-argument version of `Impl::DeepCopy` everywhere in core (i.e., passing the execution space) instead of the 3-argument version. The ultimate aim is to remove the 3-argument version.

Modified files:

- [x] `core/src/Kokkos_MemoryPool.hpp`
- [x] `core/src/impl/Kokkos_SharedAlloc_timpl.hpp`

Kudos to @masterleinad for helping me for this PR.

### CODE REVIEW

File: core/src/impl/Kokkos_SharedAlloc.hpp [hunk 1]
```diff
@@ -264,13 +264,34 @@ class SharedAllocationRecordCommon : public SharedAllocationRecord<void, void> {
...
+      ExecutionSpace{}, r_new->data(), r_old->data(),
+      std::min(r_old->size(), r_new->size()));
+  Kokkos::fence(std::string("SharedAllocationRecord<") + MemorySpace::name() +
+                ", void>::reallocate_tracked(): fence after copying data");
+
+  record_base_t::increment(r_new);
+  record_base_t::decrement(r_old);
+
+  return r_new->data();
+}
```
Comment (masterleinad) [inline core/src/impl/Kokkos_SharedAlloc.hpp:293]: The problem here was that we don't have a full definition of the execution space instance available when instantiating the class. Templating delays the instantiation until the `reallocate_tracked` member is actually needed and its only use is in `kokkos_realloc`. We could try to just inline it and get rid of this member function.

Comment (masterleinad) [reply core/src/impl/Kokkos_SharedAlloc.hpp:293]: Something like
```diff
diff --git a/core/src/Kokkos_Core.hpp b/core/src/Kokkos_Core.hpp
index 9588d289a..f6675d4aa 100644
--- a/core/src/Kokkos_Core.hpp
+++ b/core/src/Kokkos_Core.hpp
@@ -178,6 +178,21 @@ inline void* kokkos_realloc(void* arg_alloc, const size_t arg_alloc_size) {
   using MemorySpace = typename Space::memory_space;
   return Impl::SharedAllocationRecord<MemorySpace>::reallocate_tracked(
       arg_alloc, arg_alloc_size);
+
+  auto* const r_old =
+      Impl::SharedAllocationRecord<MemorySpace>::get_record(arg_alloc);
+  auto* const r_new = Impl::SharedAllocationRecord<MemorySpace>::allocate(
+      r_old->get_space(), r_old->get_label(), arg_alloc_size);
+
+  Kokkos::Impl::DeepCopy<MemorySpace, MemorySpace>(
+      r_new->data(), r_old->data(), std::min(r_old->size(), r_new->size()));
+  Kokkos::fence(std::string("SharedAllocationRecord<") + MemorySpace::name() +
+                ", void>::reallocate_tracked(): fence after copying data");
+
+  Impl::SharedAllocationRecord<void, void>::increment(r_new);
+  Impl::SharedAllocationRecord<void, void>::decrement(r_old);
+
+  return r_new->data();
 }
 
 }  // namespace Kokkos
diff --git a/core/src/impl/Kokkos_SharedAlloc.hpp b/core/src/impl/Kokkos_SharedAlloc.hpp
index e0c5b0d7a..ddfa1bf92 100644
--- a/core/src/impl/Kokkos_SharedAlloc.hpp
+++ b/core/src/impl/Kokkos_SharedAlloc.hpp
@@ -267,6 +267,7 @@ class SharedAllocationRecordCommon : public SharedAllocationRecord<void, void> {
   static void* reallocate_tracked(void* arg_alloc_ptr, size_t arg_alloc_size);
   static auto get_record(void* alloc_ptr) -> derived_t*;
   std::string get_label() const override;
+  const MemorySpace& get_space() { return m_space; }
   static void print_records(std::ostream& s, MemorySpace const&,
                             bool detail = false);
 };
@@ -331,6 +332,7 @@ class HostInaccessibleSharedAllocationRecordCommon
                             bool detail = false);
   static auto get_record(void* alloc_ptr) -> derived_t*;
   std::string get_label() const override;
+  const MemorySpace& get_space() { return m_space; }
 };
 
 #ifdef KOKKOS_ENABLE_DEBUG
```

Comment (pzehner) [reply core/src/impl/Kokkos_SharedAlloc.hpp:293]: Do you think this should be done in this PR?

Comment (masterleinad) [reply core/src/impl/Kokkos_SharedAlloc.hpp:293]: I could argue both ways. On the one hand, there are not that many other changes. On the other hand, the changes are somewhat orthogonal to the goal of this pull request. In the end, we need a good description anyway.

File: core/src/Kokkos_MemoryPool.hpp [hunk 2]
```diff
@@ -166,7 +166,8 @@ class MemoryPool {
 
     if (!accessible) {
       Kokkos::Impl::DeepCopy<Kokkos::HostSpace, base_memory_space>(
-          sb_state_array, m_sb_state_array, alloc_size);
+          typename Kokkos::HostSpace::execution_space{}, sb_state_array,
```
Comment (masterleinad) [inline core/src/Kokkos_MemoryPool.hpp:169]: Note that `base_memory_space::executiuon_space` is more likely to be able to be used in the `DeepCopy`. If the execution space can't enqueue the operation we generally fence the given execution space instance and execute it using a different execution space instance.
Also, using default constructed execution space instances might be problematic in the multi-GPU use case.

Comment (pzehner) [reply core/src/Kokkos_MemoryPool.hpp:169]: I couldn't build using `base_memory_space::execution_space` with some backends, iirc.

Comment (masterleinad) [reply core/src/Kokkos_MemoryPool.hpp:169]: All execution spaces are default-constructible. Do you mind trying again and report related problems here?

File: core/src/impl/Kokkos_SharedAlloc.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_SharedAlloc.hpp:293]: If we do this one first, we have tp justify introducing the additional dummy template argument.

File: core/src/Kokkos_MemoryPool.hpp (hunk 2 above)
Comment (pzehner) [reply core/src/Kokkos_MemoryPool.hpp:169]: Ok, this is what I get:

```
error: no instance of constructor "Kokkos::Impl::DeepCopy<Kokkos::HostSpace, MemSpace, ExecutionSpace, std::enable_if_t<<expression>, void>>::DeepCopy [with MemSpace=Kokkos::Cuda::memory_space, ExecutionSpace=Kokkos::HostSpace::execution_space]" matches the argument list
            argument types are: (Kokkos::CudaSpace::execution_space, uint32_t *const, uint32_t *const, const size_t)
        Kokkos::Impl::DeepCopy<Kokkos::HostSpace, base_memory_space>(
        ^
```

Comment (masterleinad) [reply core/src/Kokkos_MemoryPool.hpp:169]: The overload I would expect to be called is https://github.com/kokkos/kokkos/blob/7ff4042fd2dc1d1fe6ae280179fc693ad6dac6ac/core/src/Cuda/Kokkos_CudaSpace.hpp#L482-L489.
I'm surprised to see `ExecutionSpace=Kokkos::HostSpace::execution_space` and `argument types are: (Kokkos::CudaSpace::execution_space, uint32_t *const, uint32_t *const, const size_t)` at the same time in the error message. Try specifying the third template argument explicitly.

Comment (pzehner) [reply core/src/Kokkos_MemoryPool.hpp:169]: It works by passing the execution space explicitly.

File: core/src/impl/Kokkos_SharedAlloc_timpl.hpp [hunk 3]
```diff
@@ -252,11 +215,12 @@ void HostInaccessibleSharedAllocationRecordCommon<MemorySpace>::print_records(
 
   SharedAllocationHeader head;
 
+  typename MemorySpace::execution_space exec;
   if (detail) {
     do {
       if (r->m_alloc_ptr) {
         Kokkos::Impl::DeepCopy<HostSpace, MemorySpace>(
-            &head, r->m_alloc_ptr, sizeof(SharedAllocationHeader));
+            exec, &head, r->m_alloc_ptr, sizeof(SharedAllocationHeader));
```
[the remark this answers is not in the document]
Comment (pzehner) [reply core/src/impl/Kokkos_SharedAlloc_timpl.hpp:223]: What does it change exactly? Both ways seem to compile correctly.

File: core/src/Kokkos_MemoryPool.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (pzehner) [reply core/src/Kokkos_MemoryPool.hpp]: Changed for `base_memory_space::execution_space`, see previous conversation.

File: core/src/impl/Kokkos_SharedAlloc.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_SharedAlloc.hpp:293]: At the very least, add a comment explaining the additional template argument here.

File: core/src/Kokkos_MemoryPool.hpp [hunk 4]
```diff
@@ -165,8 +165,10 @@ class MemoryPool {
         accessible ? m_sb_state_array : (uint32_t *)host.allocate(alloc_size);
 
     if (!accessible) {
-      Kokkos::Impl::DeepCopy<Kokkos::HostSpace, base_memory_space>(
-          sb_state_array, m_sb_state_array, alloc_size);
+      Kokkos::Impl::DeepCopy<Kokkos::HostSpace, base_memory_space,
+                             typename base_memory_space::execution_space>(
+          typename base_memory_space::execution_space{}, sb_state_array,
```
Comment (dalg24) [inline core/src/Kokkos_MemoryPool.hpp:170]: This changes the execution space template argument.
See https://github.com/kokkos/kokkos/blob/baba8ca88410ecccf9144af18f1fa0f18203ff48/core/src/Kokkos_Core_fwd.hpp#L282-L285

Is that intentional?

Comment (pzehner) [reply core/src/Kokkos_MemoryPool.hpp:170]: This was just to make the code compile. I'll use `DstSpace::execution_space` then (i.e. `Kokkos::DefaultHostExecutionSpace`).

File: core/src/impl/Kokkos_SharedAlloc.hpp (hunk 1 above)
Comment (pzehner) [reply core/src/impl/Kokkos_SharedAlloc.hpp:293]: Ok, I'll add comments.

File: core/src/Kokkos_MemoryPool.hpp (hunk 2 above)
Comment (pzehner) [reply core/src/Kokkos_MemoryPool.hpp]: Eventually, I reverted to using `Kokkos::DefaultHostExecutionSpace`.

File: core/src/impl/Kokkos_SharedAlloc_timpl.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (JBludau) [reply core/src/impl/Kokkos_SharedAlloc_timpl.hpp:223]: nothing really, I missed line 218 :-D

File: core/unit_test/headers_self_contained/CMakeLists.txt [hunk 5]
```diff
@@ -13,6 +13,9 @@ if(NOT Kokkos_ENABLE_DEPRECATED_CODE_4 OR Kokkos_ENABLE_DEPRECATION_WARNINGS)
   list(REMOVE_ITEM KOKKOS_CORE_HEADERS "Kokkos_Future.hpp" "Kokkos_TaskScheduler.hpp")
 endif()
 
+# Remove MemoryPool from the list as it uses execution spaces
+list(REMOVE_ITEM KOKKOS_CORE_HEADERS "Kokkos_MemoryPool.hpp")
```
Comment (dalg24) [inline core/unit_test/headers_self_contained/CMakeLists.txt:17]: I do not understand.
What in these changes makes that the `<Kokkos_MemoryPool.hpp>` header is not self contained any more?

Comment (masterleinad) [reply core/unit_test/headers_self_contained/CMakeLists.txt:17]: Errors look like
```
In file included from /tmp/kokkos/core/unit_test/headers_self_contained/tstHeader.cpp:26:
In file included from tmp/kokkos/core/unit_test/headers_self_contained/tstHeader.cpp:26:
/tmp/kokkos/core/src/Kokkos_MemoryPool.hpp:169:11: error: invalid use of incomplete type 'Kokkos::DefaultHostExecutionSpace' (aka 'Kokkos::Serial')
  169 |           Kokkos::DefaultHostExecutionSpace{}, sb_state_array, m_sb_state_array,
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/kokkos/core/src/fwd/Kokkos_Fwd_SERIAL.hpp:22:7: note: forward declaration of 'Kokkos::Serial'
   22 | class Serial;  ///< Execution space main process on CPU.
      |       ^
In file included from /tmp/kokkos/core/unit_test/headers_self_contained/tstHeader.cpp:26:
/tmp/kokkos/core/src/Kokkos_MemoryPool.hpp:222:11: error: invalid use of incomplete type 'Kokkos::DefaultHostExecutionSpace' (aka 'Kokkos::Serial')
  222 |           Kokkos::DefaultHostExecutionSpace{}, sb_state_array, m_sb_state_array,
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/kokkos/core/src/fwd/Kokkos_Fwd_SERIAL.hpp:22:7: note: forward declaration of 'Kokkos::Serial'
   22 | class Serial;  ///< Execution space main process on CPU.
      |       ^
```

Comment (pzehner) [reply core/unit_test/headers_self_contained/CMakeLists.txt:17]: Do we agree on this or shall we find a different solution to make this PR pass the tests?

Comment (masterleinad) [reply core/unit_test/headers_self_contained/CMakeLists.txt:17]: I think it's fine.

File: core/unit_test/headers_self_contained/CMakeLists.txt [hunk 6]
```diff
@@ -13,6 +13,9 @@ if(NOT Kokkos_ENABLE_DEPRECATED_CODE_4 OR Kokkos_ENABLE_DEPRECATION_WARNINGS)
   list(REMOVE_ITEM KOKKOS_CORE_HEADERS "Kokkos_Future.hpp" "Kokkos_TaskScheduler.hpp")
 endif()
 
+# Remove MemoryPool from the list as it uses execution spaces
```
Comment (dalg24) [inline core/unit_test/headers_self_contained/CMakeLists.txt:16]: Change to 
```suggestion
# FIXME circular dependencies and execution space type not complete
```

**Review by dalg24 — APPROVED** [review]

**Review by masterleinad — APPROVED** [review]

**Review by JBludau — APPROVED** [review]
looks good ... we could have a FIXME on all of the comments but I am fine with how it is

**Comment by dalg24** [comment]
The CUDA runtime failure on ORNL Jenkins CI is unrelated (GPU not detected)

**Pull request merged**